Lad table vue

dtailor90 opened the pull request:
First Cut of LAD Table and LAD Table Summary views.

larkin commented on src/plugins/LADTable/LADTableSummaryViewProvider.js:
```diff
@@ -0,0 +1,68 @@
+    LADTableSummaryComponent,
+    Vue
+) {
+    function Tabs(openmct) {
```
Tabs?

larkin commented on src/plugins/LADTable/LADTableViewProvider.js:
```diff
@@ -0,0 +1,68 @@
+    LADTableComponent,
+    Vue
+) {
+    function Tabs(openmct) {
```
`Tabs`?

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,98 @@
+
+<template>
+    <tr>
+        <td>{{dObject.name}}</td>
```
just call it domainObject

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,98 @@
+    props: ['dObject'],
+    data() {
+        return {
+            name: this.dObject.name,
```
unused.

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,98 @@
+            .observe(
+                this.dObject,
+                '*',
+                this.updateName
```
doesn't exist.

larkin commented on src/plugins/LADTable/components/LADTable.vue:
```diff
@@ -0,0 +1,75 @@
+            </tr>
+        </thead>
+        <tbody>
+            <LADRow 
```
component names should be kebab case, i.e. `lad-row`

larkin commented on src/plugins/LADTable/components/LADTable.vue:
```diff
@@ -0,0 +1,75 @@
+ import LADRow from './LADRow.vue';
+
+ export default {
+    inject: ['openmct', 'domainObject', 'composition'],
```
simpler to just set composition in `loaded` instead of `openmct`.

larkin commented on src/plugins/LADTable/LADTableSummaryViewProvider.js:
```diff
@@ -0,0 +1,68 @@
+                            provide: {
+                                openmct,
+                                domainObject,
+                                composition: openmct.composition.get(domainObject)
```
We use injectors in the plugin for injecting openmct and domainObject, and I expect we will find a better way to do this in the future.  In this case, you're also passing composition via injection, and it would be more consistent with the rest of the app if you set composition in `data` method or `mounted` function of the component.

larkin commented on src/plugins/LADTable/plugin.js:
```diff
@@ -0,0 +1,56 @@
+            });
+
+            openmct.types.addType('LADTableSummary', {
+                name: "LAD Table Summary",
```
How about "LAD Table Set" instead of "LAD Table Summary"?  As in, "a set of LAD tables".

The idea is that a lad table set should include a separator (the name of the lad table) between each set of channels).

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,101 @@
+                this.updateName
+            );
+
+        this.openmct.time.on('timeSystem', this.updateTimeSystem, this);
```
doesn't exist.

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,101 @@
+
+        this.unsubscribe = this.openmct
+            .telemetry
+            .subscribe(this.domainObject, this.updateValues.bind(this), {});
```
vue methods don't need to be bound.  and what's the third parameter?

larkin commented on src/plugins/LADTable/components/LadTableSet.vue:
```diff
@@ -0,0 +1,104 @@
+                v-for="primary in primaryCollection">
+
+                <tr
+                    :key="primary.identifier.key">
```
Keys are not unique on their own.

larkin commented on src/plugins/LADTable/components/LADTable.vue:
```diff
@@ -0,0 +1,76 @@
+        <tbody>
+            <lad-row 
+                v-for="item in items"
+                :key="item.identifier.key"
```
identifier.key on it's own is not unique.  you need to make a keystring.

larkin commented on src/plugins/LADTable/components/LADTable.vue:
```diff
@@ -0,0 +1,76 @@
+    },
+    mounted() {
+        this.composition = this.openmct.composition.get(this.domainObject);
+        this.composition.on('add', this.addItem);
```
need to handle remove.

dtailor90 replied:
handled in latest commits

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,101 @@
+            this.value = this.formats[this.valueKey].format(datum);
+            var limit = this.limitEvaluator.evaluate(datum, this.valueKey);
+            if (limit) {
+                this.valueClass = limit.cssClass;
```
not doing anything with value class (and you should be)

dtailor90 replied:
i have added the class to the cell

dtailor90 commented:
@larkin - Made changes. I have added logic to handle composition remove, on both tables and sets, added the valueClass to the value cell, and updated Time System on time system change. Also removed unnecessary code. Please re-review

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,111 @@
+    <tr>
+        <td>{{name}}</td>
+        <td>{{timestamp}}</td>
+        <td :class="[valueClass]">
```
This doesn't need to be wrapped in an array.

larkin commented on src/plugins/LADTable/LADTableViewProvider.js:
```diff
@@ -0,0 +1,67 @@
+    LadTableComponent,
+    Vue
+) {
+    function LADTableView(openmct) {
```
classname should match filename.

larkin commented on src/plugins/LADTable/components/LADTable.vue:
```diff
@@ -0,0 +1,94 @@
+ </template>
+
+ <script>
+ import LadRow from './LadRow.vue';
```
not every file system is case-insensitive, this import needs to match filename.

larkin commented on src/plugins/LADTable/components/LADTable.vue:
```diff
@@ -0,0 +1,94 @@
+            let index = -1;
+
+            array.forEach((o,i) => {
+                if(o.domainObject.identifier.key === object.key) {
```
identifier cannot be compared by key alone.  

also, this would be a good case to use `_.findIndexBy`.

larkin commented on src/plugins/LADTable/components/LadTableSet.vue:
```diff
@@ -0,0 +1,146 @@
+    },
+    data() {
+        return {
+            primaryCollection: [],
```
can we name these more clearly?  it's not clear what they represent or how they relate to a "LAD table"

larkin commented on src/plugins/LADTable/plugin.js:
```diff
@@ -0,0 +1,56 @@
+                }
+            });
+
+            openmct.types.addType('LadTableSet', {
```
type casing should be consistent.

dtailor90 replied:
Sorry, made the change

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,111 @@
+        updateName(name){
+            this.name = name;
+        },
+        updateTimeSystem(timeSystem) {
```
changing timesystem should clear display and re-request latest value.

dtailor90 replied:
Done

larkin commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -0,0 +1,111 @@
+        this.openmct
+            .telemetry
+            .request(this.domainObject, {strategy: 'latest'})
+            .then((values) => values.forEach(this.updateValues));
```
considering this can only display a single value, this should only call updateValues once instead of iterating over every value in the response.
